server/src/main/java/org/opensearch/index/remote/RemoteMigrationClusterStateUtils.java

```diff
@@ -0,0 +1,191 @@
+     * @param discoveryNodes       set of discovery nodes from cluster state
+     * @param currentIndexSettings current {@link IndexMetadata} from cluster state
+     * @return <code>true</code> or <code>false</code> depending on the met conditions
+     */
+    public static boolean needsRemoteIndexSettingsUpdate(
+        IndexRoutingTable indexRoutingTable,
+        DiscoveryNodes discoveryNodes,
+        Settings currentIndexSettings
+    ) {
+        assert currentIndexSettings != null : "IndexMetadata for a shard cannot be null";
+        if (indexHasRemoteStoreSettings(currentIndexSettings) == false) {
+            boolean allPrimariesStartedAndOnRemote = indexRoutingTable.shardsMatchingPredicate(ShardRouting::primary)
+                .stream()
+                .allMatch(shardRouting -> shardRouting.started() && discoveryNodes.get(shardRouting.currentNodeId()).isRemoteStoreNode());
+            List<ShardRouting> replicaShards = indexRoutingTable.shardsMatchingPredicate(shardRouting -> shardRouting.primary() == false);
+            boolean noRelocatingReplicas = replicaShards.stream().noneMatch(ShardRouting::relocating);
+            boolean allStartedReplicasOnRemote = replicaShards.stream()
+                .filter(ShardRouting::started)
+                .allMatch(shardRouting -> discoveryNodes.get(shardRouting.currentNodeId()).isRemoteStoreNode());
+            return allPrimariesStartedAndOnRemote && noRelocatingReplicas && allStartedReplicasOnRemote;
+        }
+        return false;
+    }
+
+    /**
+     * Updates the remote store path strategy metadata for the index when it is migrating to remote.
+     * This should be run only when the first primary copy moves over from docrep to remote.
+     * Checks are in place to make this execution no-op if the index metadata is already present
+     *
+     * @param indexMetadata        Current {@link IndexMetadata}
+     * @param indexMetadataBuilder Mutated {@link IndexMetadata.Builder} having the previous state updates
+     * @param index                index name
+     * @param discoveryNodes       Current {@link DiscoveryNodes} from the cluster state
+     * @param settings             current cluster settings from {@link ClusterState}
+     */
+    public static void maybeUpdateRemoteStorePathStrategy(
+        IndexMetadata indexMetadata,
+        IndexMetadata.Builder indexMetadataBuilder,
+        String index,
+        DiscoveryNodes discoveryNodes,
+        Settings settings
+    ) {
+        if (indexHasRemotePathMetadata(indexMetadata) == false) {
+            logger.info("Adding remote store path strategy for index [{}] during migration", index);
+            indexMetadataBuilder.putCustom(REMOTE_STORE_CUSTOM_KEY, createRemoteStorePathTypeMetadata(settings, discoveryNodes));
+        } else {
+            logger.debug("Does not match criteria to update remote store path type for index {}", index);
```

Comment:
may be say : "index already have remote store path type"

Reply:
Done